Restore the Deltas! GoalVC includes deltas again, now sourced from the dueby

krugerk opened the pull request:
## Summary
Includes the due by table in the goal, displaying the deltas on the goal screen, where delta_text was previously displayed.
The user can quickly see, at the goal screen, how many of a thing to do to obtain a number of days of safety.

*For UI changes including screenshots of before and after is great.*
![due by deltas](https://github.com/user-attachments/assets/3900e32a-d902-4b95-873a-09b91973e0ed)

and now colored
![due by deltas, colored](https://github.com/user-attachments/assets/fe97ffb0-c0d7-4e4a-b8e1-1f9b9f56f91b)
![due by deltas, colored, on light](https://github.com/user-attachments/assets/b8e8db49-bd07-47f1-a103-dda1124cba16)



## Validation
built and ran the app in the simulator
viewed the goal screen of various goals


Relates to #209 
This shows the deltas on the goal screen. It does not display the entire table and it does not show the totals nor does it show the dates. It restores "amounts to do" previously provided by "delta text". (See #455)
Fixes #559

krugerk committed:
Introduce DueBy Table Model, show deltas on goal's screen

previously "delta text" was shown in the goalvc, which provided
insight into the number of things due in the next days

storing the dictionary with a custom type inside core data as a (json) string
... as opposed to making separate entities

at least partially
resolves #209
fixes #559

krugerk commented on BeeKit/Model/Goal.swift:
```diff
@@ -184,3 +201,113 @@ public class Goal: NSManagedObject {
+                self.dueByTableData = jsonString
+            }
+        } catch {
+            print("Error encoding dictionary: \(error)")
```
logging instead of mere printing

krugerk committed:
clean code

krugerk commented on BeeSwift/GoalViewController.swift:
```diff
@@ -488,7 +501,15 @@ class GoalViewController: UIViewController,  UIScrollViewDelegate, DatapointTabl
+        self.dueByLabel.text = goal.dueBy
+            .sorted(using: SortDescriptor(\.key))
+            .compactMap { $0.value.formatted_delta_for_beedroid }
+            .map { $0 == "✔" ? "✓" : $0 }
```
preferring this other checkmark, which shows up well on both light and dark

It is also the same used to indicate a goal has data for the day, today (when "todayta" is true, see https://github.com/beeminder/BeeSwift/blob/745cf9875ed6b0303c46d5d8e4c0a2d44796040d/BeeSwift/GoalCollectionViewCell.swift#L27)

krugerk committed:
coloring due by deltas on the goalvc

uses the same general coloring as the website does:
orange for the amount to do today to be safe til tomorrow
blue to become safe for two days
green for three days of safety
and the plain text color after that

dreeves commented:
I don't love the aesthetics of this. Maybe also in terms of user-experience, it's cluttery and overwhelming and opaque.

krugerk commented:
> I don't love the aesthetics of this. Maybe also in terms of user-experience, it's cluttery and overwhelming and opaque.

The expectation is that the goal screen contains an expandable section, maybe just a label "due by table" where upon tapping it, it expands to show the full table? It seems you may have suggested as much previously https://github.com/beeminder/BeeSwift/issues/36#issuecomment-627741771 .

theospears commented on BeeKit/Model/Goal.swift:
```diff
@@ -62,6 +62,14 @@ public class Goal: NSManagedObject {
+    
+    public var dueBy: [String: BeeminderDueByEntry] {
+        self.loadDueByTable() ?? [:]
+    }
```
Can we not use a [ValueTransformer](UICollectionViewDiffableDataSource) to avoid maintaining both properties?

krugerk replied:
I think you meant to link to something else with the ValueTransformer link there. I wanted to avoid making a separate core data entity for the due by table. I can look into the value transformer.

theospears replied:
You're right, I was trying to link to https://developer.apple.com/documentation/coredata/handling_different_data_types_in_core_data#3582174

Although https://developer.apple.com/documentation/foundation/valuetransformer might have been a better link

krugerk replied:
done

theospears commented:
Sharing my higher level thoughts here
* We clearly need to provide this date (or at least a subset of it) to users in the app
* I'm not convinced this is the right UI
* But once the data model parts of this PR are in a good place, it seems reasonable that we merge it to be a starting point to iterate the UI on.

krugerk commented:
> I'm not convinced this is the right UI

For all of the users of the app on iOS, as is, the deltas shown basically look like they did when the users knew them from delta_text. So any user on iOS should be automatically familiar with this interface.

The MR is here as a "request for comments/feedback/suggestions/...". May its spark encourage more and may it evolve from creation til merge.

krugerk committed:
same font as before with delta_text

krugerk committed:
same safety buffer colors

krugerk committed:
clean code

krugerk committed:
use ValueTransformer for DueByTable in Core Data

krugerk committed:
unused parameter removed from init

krugerk committed:
clean code

theospears commented on BeeSwift/GoalViewController.swift:
```diff
@@ -515,7 +528,10 @@ class GoalViewController: UIViewController,  UIScrollViewDelegate, DatapointTabl
         self.datapointTableController.datapoints = goal.recentData.sorted(by: {$0.updatedAt < $1.updatedAt})
-
+        
+        self.deltasLabel.isHidden = goal.dueByTable.entries.isEmpty
```
[Nit] This won't collapse the height - maybe an argument for StackView here as well. Doesn't seem like a big deal though.

krugerk replied:
I would hope the app's UI would all be written in `SwiftUI` before long. Shall the views first be migrated to `UIKit`'s `UIStackView`?

theospears commented on BeeKit/Model/Goal.swift:
```diff
@@ -62,6 +62,11 @@ public class Goal: NSManagedObject {
     @objc(removeRecentData:)
     @NSManaged public func removeFromRecentData(_ values: Set<DataPoint>)
+    
+    @NSManaged public var dueByTable: DueByTable
```
I suggest calling this variable `dueBy` - I've been trying to exactly match api names (modulo casing and underscores) so I don't need to remember both.

theospears commented on BeeKit/Model/DueBy/DueByTable.swift:
```diff
@@ -0,0 +1,38 @@
+        var entries: [String : BeeminderDueByEntry] {
+            dueByJson?.dictionary?.compactMapValues(BeeminderDueByEntry.init)
+            ?? dueByJson?.dictionary?.mapValues(BeeminderDueByEntry.init)
+            ?? [:]
```
Could you explain what is happening here?

theospears commented on BeeKit/Model/DueBy/DueByTable.swift:
```diff
@@ -0,0 +1,38 @@
+        self.entries = entries
+    }
+    
+    public static var supportsSecureCoding: Bool { true }
```
From: https://developer.apple.com/documentation/foundation/nssecurecoding

> An object that does override [init(coder:)](https://developer.apple.com/documentation/mapkit/mkmapitem/init(coder:)) must decode any enclosed objects using the [decodeObjectOfClass:forKey:](https://developer.apple.com/documentation/foundation/nscoder/1442558-decodeobjectofclass) method. For example:

It appears we aren't doing this, so shouldn't be returning true here?

krugerk replied:
True. It seems I missed one entry.

theospears commented on BeeKit/Model/DueBy/BeeminderDueByEntry.swift:
```diff
@@ -0,0 +1,56 @@
+        case formattedDeltaForBeedroid
+    }
+    
+    public func encode(with coder: NSCoder) {
```
Do we need all this custom encode/decode logic? My understanding is Swift (4.0+) should be able to synthesize it.

krugerk replied:
Codable can take care of some of it. The rest is for NSCoding (NSSecureCoding). We can make one use the other, though this is not available out of the box, afaik. Will look into it

theospears replied:
If we are using NSSecureCoding, can we use `NSSecureUnarchiveFromDataTransformer` instead of a custom one?

krugerk replied:
I do not think so. I think if custom, core data non-native, are being stored, then a custom valuetransformer is needed.

krugerk replied:
> Do we need all this custom encode/decode logic?

This uses the pattern the rest of the app uses: SwiftyJSON and manually plucking out from the json.

> My understanding is Swift (4.0+) should be able to synthesize it.

NSCoding's encode: `public func encode(with coder: NSCoder)`
Codable's encode: `public func encode(to encoder: Encoder) throws`

theospears reviewed (changes requested):
Thanks for updating this to ValueTransformer. 
The most important bit of feedback here is: Do we actually need to have all the custom coding/decoding logic, or can the swift compiler generate it for us?
https://developer.apple.com/documentation/foundation/archives_and_serialization/encoding_and_decoding_custom_types#2904055

krugerk committed:
decodeObject(of:forKey) for NSSecureCoding

krugerk commented on BeeSwift/GoalViewController.swift:
```diff
@@ -656,3 +672,42 @@ private extension GoalViewController {
+
+
+private extension GoalViewController {
+    var dueByContainsSpecificAmounts: Bool {
```
It is. The idea was just to show the deltas when at least one of them was a number (not just all checkmarks). I am thinking it might be better to just leave the checkmarks and thus this code remains unused and can be removed.

krugerk commented on BeeSwift/AppDelegate.swift:
```diff
@@ -31,6 +31,8 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
             UIFont.beeminder.defaultFontPlain.withSize(20)]
         UIBarButtonItem.appearance().setTitleTextAttributes([NSAttributedString.Key.font : UIFont.beeminder.defaultFontPlain.withSize(18)], for: UIControl.State())
 
+        DueByTableValueTransformer.register()
```
Yes, this makes sense - move it to the location where the rest of core data is initialized.

krugerk committed:
custom value transformer was not being registered close to where core data was being initialized

krugerk committed:
remove cruft

krugerk committed:
swift codable for DueByTable

krugerk committed:
remove cruft

krugerk committed:
debugging can be easier with a description

krugerk committed:
clean code

krugerk committed:
redundant public

theospears commented:
I put together an example of what I was thinking with regards to using Codable in #617.

krugerk committed:
goal contains dueby dictionary

krugerk committed:
clean code

krugerk committed:
CodingKeys of CodingKey

krugerk committed:
clean code

krugerk commented:
> I put together an example of what I was thinking with regards to using Codable in #617.

OK. This merge request has been updated accordingly: removed the intermediary container that just stored the dictionary. Removed the NS(Secure)Coding of `BeeminderDueByEntry`. Used lowerCamelCase instead of snake_case.

krugerk commented on BeeKit/Model/Goal.swift:
```diff
@@ -62,6 +62,8 @@ public class Goal: NSManagedObject {
     @objc(removeRecentData:)
     @NSManaged public func removeFromRecentData(_ values: Set<DataPoint>)
+    
+    @NSManaged public var dueByDaystamp: DueByDictionary
```
```suggestion
    @NSManaged public var dueByDay: DueByDictionary
```

```suggestion
    @NSManaged public var dueByDayMap: DueByDictionary
```

```suggestion
    @NSManaged public var dueByDate: DueByDictionary
```

krugerk committed:
clean code

krugerk committed:
clean code

krugerk committed:
Merge branch 'master' into feature/209-due-by-table
